pallets/automation-time/src/lib.rs

```diff
@@ -1343,27 +1325,74 @@ pub mod pallet {
 			Ok(())
 		}
 
-		/// Reschedules an existing task for a given number of execution times
-		fn reschedule_existing_task(
+		fn reschedule_or_remove_task(
 			task_id: TaskId<T>,
-			task: &TaskOf<T>,
-			execution_times: Vec<UnixTime>,
-		) -> Result<(), DispatchError> {
-			let new_executions = execution_times.len().try_into().unwrap();
+			task: &mut TaskOf<T>,
+			dispatch_error: Option<DispatchError>,
+		) {
+			let remove_on_err = |task: &TaskOf<T>, err: DispatchError| {
+				Self::deposit_event(Event::<T>::TaskNotRescheduled {
+					who: task.owner_id.clone(),
+					task_id,
+					error: err,
+				});
+				AccountTasks::<T>::remove(task.owner_id.clone(), task_id);
+			};
 
-			let fee = Self::calculate_execution_fee(&task.action, new_executions)?;
-			T::FeeHandler::can_pay_fee(&task.owner_id, fee.clone())
-				.map_err(|_| Error::<T>::InsufficientBalance)?;
+			if let Some(e) = dispatch_error {
+				remove_on_err(task, e);
+			} else {
+				let owner_id = task.owner_id.clone();
+				match Self::reschedule_existing_task(task_id, task) {
+					Ok(_) => {
+						Self::deposit_event(Event::<T>::TaskRescheduled { who: owner_id, task_id });
+					},
+					Err(e) => {
+						remove_on_err(task, e);
+					},
+				};
+			}
+		}
 
-			Self::insert_scheduled_tasks(task_id, task, execution_times.clone())?;
+		fn reschedule_existing_task(task_id: TaskId<T>, task: &mut TaskOf<T>) -> DispatchResult {
+			match task.schedule {
+				Schedule::Recurring { ref mut next_execution_time, frequency } => {
+					let new_execution_time = next_execution_time
+						.checked_add(frequency)
+						.ok_or(Error::<T>::InvalidTime)?;
+					*next_execution_time = new_execution_time;
 
-			T::FeeHandler::withdraw_fee(&task.owner_id, fee.clone())
-				.map_err(|_| Error::<T>::LiquidityRestrictions)?;
+					// TODO: should execution fee depend on whether task is recurring?
+					let fee = Self::calculate_execution_fee(&task.action, 1)?;
```

Comment:
this doesn't take the XCMP fees into account

Reply:
I don't have a good grasp on xcmp fees yet. How are you thinking that I should include the xcmp fees here?

Reply:
we just need to collect it here. The xcmp fees are separate from the execution. Right now this will cause someone's xcmp action to fail because there won't be enough funds in the account to pay for it. You can see how this is done in `validate_and_schedule_task` which you already updated. Might be worth trying to refactor to use the same fee logic.

Additionally I am now agreeing that we should factor this weight into the execution fee, but that can come later.

Reply:
Rather than make this pr larger, I broke the structural changes to fee payment into its own pr: https://github.com/OAK-Foundation/OAK-blockchain/pull/298
Will wait on merging the fee pr before moving forward with this.